File: 프로그래머스/unrated/250137. ［PCCP 기출문제］ 1번 ／ 붕대 감기/［PCCP 기출문제］ 1번 ／ 붕대 감기.js
function solution(bandage, health, attacks) {
    let answer = 0
    let attack_dict = new Map()
    for(var i=0; i<attacks.length; i++){
        attack_dict.set(attacks[i][0],attacks[i][1])
    }
    
    let t = 0 
    let c_time = 0
    let max_time = attacks[attacks.length-1][0]
    let max_health = health
    
    while(t<=max_time){
        // t가 attack_dict에 있을때
        if(attack_dict.has(t)){
            health -= attack_dict.get(t)
            c_time = 0
            if(health<=0){
                return -1
            }
        } else {
            c_time += 1
            if(c_time < bandage[0]) {
                health += bandage[1]
                if(health > max_health){
                    health = max_health
                }
            } else {
                health += bandage[1] + bandage[2]
                if(health>max_health){
                    health = max_health
                }
                c_time = 0
            }
        }
        t += 1
    }
    return health;
}

File: 프로그래머스/unrated/250137. ［PCCP 기출문제］ 1번 ／ 붕대 감기/［PCCP 기출문제］ 1번 ／ 붕대 감기.py
def solution(bandage, health, attacks):
    answer = 0
    # 공격 정보 
    attack_dict = {}
    for attack,damage in attacks:
        attack_dict[attack] = damage
    
    t = 0
    consecutive = 0 # 연속 성공
    max_time = attacks[-1][0]
    max_health = health # 유지해야하는 체력 양 
    
    while t <= max_time:
        # 공격 time이 왔을때 
        if t in attack_dict:
            health -= attack_dict[t]
            consecutive = 0
            
            if health <= 0:
                return -1
        
        else:
            consecutive += 1
            # 연속시간이 시전시간에 도달하지 않았을때: 
            if consecutive < bandage[0]:
                health += bandage[1]
                if health > max_health:
                    health = max_health
            # 연속시간이 시전시간을 넘었을때 
            else:
                health += bandage[1] + bandage[2]
                if health > max_health:
                    health = max_health
                consecutive = 0 
        t += 1
        
    return health